sklearn/utils/validation.py

```diff
@@ -568,6 +568,22 @@ def check_array(array, accept_sparse=False, *, accept_large_sparse=True,
     if hasattr(array, 'sparse') and array.ndim > 1:
         # DataFrame.sparse only supports `to_coo`
         array = array.sparse.to_coo()
+        if array.dtype == np.dtype('object'):
+            def is_sparse_df_with_mixed_types(df):
+                dtps = set([dt.subtype.name for dt in df.dtypes])
+                return len(dtps) > 1
+
+            mixed = is_sparse_df_with_mixed_types(array_orig)
+            mxg = "A coo_matrix with dtype 'np.object' is created "\
+                  "when \nDataFrame has sparse extention arrays of"\
+                  " mixed numeric types."\
+                if mixed else \
+                "SparseArrays of dtype: bytes, str or object lead"\
+                "to the generation of a coo_matrix of dtype object"
+            raise ValueError(
+                "Pandas DataFrame with sparse extention arrays generated "
+                "a sparse coo_matrix with dtype 'np.object'\n which "
+                "scipy cannot convert to a CSR or a CSC. {}".format(mxg))
```

Comment:
For now, I do not think we need to be specific about the cause. I think a message such as:

> Pandas DataFrame with mixed sparse extension arrays generated a sparse matrix with `object` dtype which can not be converted to a scipy sparse matrix

Reply:
hey @thomasjpfan. There are two scenarios, that can result in scipy coo_matrix.to_csr throwing the Exception:
`TypeError: no supported conversion for types: (dtype('O'),)`
a. check_array receives a DataFrame with columns of different numeric types
b. check_array receives a data frame where all the columns are the same type but this type is str or bytes or object

```
tf = pd.DataFrame({'col1': pd.arrays.SparseArray([0, 1, 0],
                                                             dtype='str'),
                   'col2': pd.arrays.SparseArray([1, 0, 1],
                                                             dtype='str')})
a = check_array(tf, **{'accept_sparse': ['csr', 'csc'],
                'ensure_min_features': 2})
...
raise TypeError('no supported conversion for types: %r' % (args,))
TypeError: no supported conversion for types: (dtype('O'),)
```
In this fix I try to handle both cases since they both throw the same scipy exception.

For <case b>, I don't think we can display the message:
`Pandas DataFrame with mixed sparse extension arrays generated a sparse matrix with object dtype which can not be converted to a scipy sparse matrix`.

So you propose that we won't handle <case b>, meaning that we let the system throw the original exception?
`TypeError: no supported conversion for types: (dtype('O'),)`  ?

Reply:
> So you propose that we won't handle scenario b, meaning that we let the system throw the original exception?

Yes.

I do not think `check_array` should expect SparseArrays with strings or objects. Raising an error for the common use case of mixed numerical data as specified in https://github.com/scikit-learn/scikit-learn/issues/17945 would be enough.

I am trying to avoid too much pandas specific code in the test suite. I think a simple test for mixed numerical sparse arrays, would be sufficient to resolve the original issue.

Reply:
Got it. Will do. Thanks a lot.